### PR TITLE
Feat:[PKD-003][PKD-006][PKD-020]Pokemon Information Page: Moves & Fix…

### PR DESCRIPTION
**Pull request by Enrickzz**
… Get Pokemon Type

- details_pokemon_handler update to get pokemon moves
- Get Pokemon Type on PostFrameCallback
- base_stats ui updates
- pokemon details page UI updates (colors)
- added pokemon_moves_tab widget (UI and Implementation)
- added universal_functions.dart and color constants in constants.dart

https://user-images.githubusercontent.com/61441278/181159752-eeaa6df5-b9e8-49ce-b9b4-490567ba08c0.mp4

### CODE REVIEW

File: lib/feature/home_page/widgets/pokemon_tile.dart [hunk 1]
```diff
@@ -7,6 +7,7 @@ import 'package:pokedex_async_redux/feature/pokemon_details_page.dart/pokemon_de
 import 'package:pokedex_async_redux/utilities/constants.dart';
 import 'package:flutter/material.dart';
 import 'package:http/http.dart' as http;
+import 'package:pokedex_async_redux/utilities/universal_functions.dart';
 
 class PokemonTile extends StatefulWidget {
   const PokemonTile({
```
Comment (dbenedictos) [inline lib/feature/home_page/widgets/pokemon_tile.dart]: trailing comma

File: lib/feature/home_page/widgets/pokemon_tile.dart [hunk 2]
```diff
@@ -22,25 +23,30 @@ class _PokemonTileState extends State<PokemonTile> {
...
-          thisTileTypes = PokemonType(
-            id: result['id'],
-            subTypes: SubTypeMap.map((e) => SubType(
-                slot: result['slot'],
-                type: SpecificType(
-                  name: e['type']['name'],
-                  url: e['type']['url'],
-                ))).toList(),
-          );
+    if (mounted) {
```
Comment (dbenedictos) [inline lib/feature/home_page/widgets/pokemon_tile.dart]: put the mounted check inside the post frame callback. Reason is that if this holds true, and then, for some reason, the widget suddenly has been unmounted after the mounted check, everything inside the callback will still be called.

File: lib/feature/pokemon_details_page.dart/pokemon_details_page.dart [hunk 3]
```diff
@@ -58,13 +62,13 @@ class PokemonDetails extends StatelessWidget {
                                   Container(
                                     margin: EdgeInsets.fromLTRB(0, 0, 5, 0),
                                     decoration: BoxDecoration(
-                                      color: Colors.grey, // TODO: Make dynamic later
+                                      color: getPokemonColor('typeDetailsPage'), // TODO: Make dynamic later
```
Comment (dbenedictos) [inline lib/feature/pokemon_details_page.dart/pokemon_details_page.dart]: strings file

File: lib/feature/pokemon_details_page.dart/pokemon_details_page.dart [hunk 4]
```diff
@@ -78,13 +82,13 @@ class PokemonDetails extends StatelessWidget {
                                         : true,
                                     child: Container(
                                       decoration: BoxDecoration(
-                                        color: Colors.grey, // change color later
+                                        color: getPokemonColor('typeDetailsPage'),
```
Comment (dbenedictos) [inline lib/feature/pokemon_details_page.dart/pokemon_details_page.dart]: string files

File: lib/feature/pokemon_details_page.dart/pokemon_details_page.dart [hunk 5]
```diff
@@ -156,10 +163,12 @@ class PokemonDetails extends StatelessWidget {
...
                                     children: <Widget>[
-                                      Container(color: Colors.amberAccent),
+                                      Container(color: Colors.white),
                                       BaseStats(stat: pokemon.stats!),
-                                      Container(color: Colors.amberAccent),
-                                      Container(color: Colors.amberAccent),
+                                      Container(color: Colors.white),
+                                      MovesTab(
+                                          moves: pokemon.moves,
+                                          color: getPokemonColor(pokemon.types?.first.type?.name ?? '')),
```
Comment (dbenedictos) [inline lib/feature/pokemon_details_page.dart/pokemon_details_page.dart]: trailing comma

File: lib/utilities/universal_functions.dart [hunk 6]
```diff
@@ -0,0 +1,55 @@
+import 'package:flutter/material.dart';
+import 'package:pokedex_async_redux/utilities/constants.dart';
+
+String capitalizeFirstLetter(String name) {
```
Comment (dbenedictos) [inline lib/utilities/universal_functions.dart]: instead of a function, make an extension instead.

File: lib/utilities/universal_functions.dart [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+      return kPoison;
+    case 'psychic':
+      return kPsychic;
+    case 'rock':
+      return kRock;
+    case 'steel':
+      return kSteel;
+    case 'water':
+      return kWater;
+    case 'typeDetailsPage':
```
Comment (dbenedictos) [inline lib/utilities/universal_functions.dart]: what is `typeDetailsPage`?

File: lib/utilities/universal_functions.dart [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+
+String capitalizeFirstLetter(String name) {
+  if (name != '') {
+    return '${name[0].toUpperCase()}${name.substring(1)}';
+  } else {
+    return '';
+  }
+}
+
+Color getPokemonColor(String type) {
```
Comment (dbenedictos) [inline lib/utilities/universal_functions.dart]: I think this can be an extension too.

**Review by dbenedictos — CHANGES_REQUESTED** [review]

File: lib/utilities/universal_functions.dart (hunk 7 above)
Comment (Enrickzz) [reply lib/utilities/universal_functions.dart]: to get the color for this widget
image

File: lib/feature/pokemon_details_page.dart/pokemon_details_page.dart [hunk 9]
```diff
@@ -78,13 +82,13 @@ class PokemonDetails extends StatelessWidget {
                                         : true,
                                     child: Container(
```
Comment (dbenedictos) [inline lib/feature/pokemon_details_page.dart/pokemon_details_page.dart]: Make a separate widget for this, so you can just reuse the widget wherever.

**Review by dbenedictos — CHANGES_REQUESTED** [review]

File: lib/utilities/universal_functions.dart (hunk 6 above)
Comment (dbenedictos) [reply lib/utilities/universal_functions.dart]: when you're done, put the extension inside the extensions.dart file inside your utilities folder

**Review by dbenedictos — CHANGES_REQUESTED** [review]

File: lib/feature/pokemon_details_page.dart/pokemon_details_page.dart (hunk 9 above)
Comment (dbenedictos) [reply lib/feature/pokemon_details_page.dart/pokemon_details_page.dart]: line 80: 'pokemon.types?.last.type?.name == pokemon.types?.first.type?.name' already returns a boolean, so why make is a ternary?

**Review by dbenedictos — CHANGES_REQUESTED** [review]

File: lib/utilities/universal_functions.dart (hunk 7 above)
Comment (dbenedictos) [reply lib/utilities/universal_functions.dart]: do you use the white colour? If not make this the default instead

**Review by dbenedictos — CHANGES_REQUESTED** [review]

File: lib/feature/home_page/widgets/pokemon_tile.dart (hunk 2 above)
Comment (dbenedictos) [reply lib/feature/home_page/widgets/pokemon_tile.dart]: Screenshot 2022-07-27 at 1 38 08 PM
for example. Still encountered the error 😛

**Review by dbenedictos — CHANGES_REQUESTED** [review]

File: lib/utilities/extensions.dart [hunk 10]
```diff
@@ -0,0 +1,48 @@
+import 'package:flutter/material.dart';
+import 'package:pokedex_async_redux/utilities/constants.dart';
+
+extension StringExt on String {
```
Comment (dbenedictos) [inline lib/utilities/extensions.dart:4]: noice

File: lib/feature/home_page/widgets/pokemon_tile.dart [hunk 11]
```diff
@@ -7,26 +7,25 @@ import 'package:pokedex_async_redux/feature/pokemon_details_page.dart/pokemon_de
...
-  PokemonType? thisTileTypes;
+  PokemonType? thisTileTypes = PokemonType();
   @override
   void initState() {
-    http.get(Uri.tryParse('${widget.thisPokemon.url}') ?? Uri()).then((value) {
-      if (value.statusCode == 200) {
-        var result = jsonDecode(value.body);
-        List SubTypeMap = result['types'];
-        setState(() {
+    WidgetsBinding.instance.addPostFrameCallback((_) {
```
Comment (dbenedictos) [inline lib/feature/home_page/widgets/pokemon_tile.dart:24]: postframe((){
   if (mounted){
       doSomething();
   }
})

File: lib/feature/home_page/widgets/pokemon_tile.dart [hunk 12]
```diff
@@ -36,14 +35,23 @@ class _PokemonTileState extends State<PokemonTile> {
...
+          setState(() {});
+        } else {
+          print('error');
+        }
+      });
     });
     super.initState();
   }
 
+  @override
```
Comment (dbenedictos) [inline lib/feature/home_page/widgets/pokemon_tile.dart:50]: no need

File: lib/utilities/constants.dart [hunk 13]
```diff
@@ -1,6 +1,29 @@
...
+const Color kPsychic = Color(0xfff35a93);
+const Color kRock = Color(0xff948352);
+const Color kIce = Color(0xff6bd1e4);
+const Color kGhost = Color(0xff8d58a2);
+const Color kDragon = Color(0xffd2c0f8);
+const Color kDark = Color(0xff4f4f4f);
+const Color kSteel = Color(0xff7c8e8a);
+const Color kFairy = Color(0xfffda1c2);
+
+const Color bgTypeDetailsPage = Color.fromARGB(99, 255, 255, 255);
```
Comment (dbenedictos) [inline lib/utilities/constants.dart]: `typeDetailsPageBackgroundColor`

File: lib/feature/pokemon_details_page.dart/pokemon_details_page.dart [hunk 14]
```diff
@@ -14,13 +17,13 @@ class PokemonDetails extends StatelessWidget {
         slivers: <Widget>[
           SliverAppBar(
             leading: BackButton(color: Colors.black),
-            backgroundColor: Colors.white, // TODO: Make dynamic later
+            backgroundColor: pokemon.types?.first.type?.name?.getPokemonColor ?? ''.getPokemonColor,
```
Comment (dbenedictos) [inline lib/feature/pokemon_details_page.dart/pokemon_details_page.dart]: for readability
`(pokemon.types?.first.type?.name?.getPokemonColor ?? '').getPokemonColor`
and do the same for the rest

File: lib/utilities/extensions.dart [hunk 15]
```diff
@@ -0,0 +1,48 @@
...
+  String get capitalize => '${this[0].toUpperCase()}${substring(1)}';
```
Comment (dbenedictos) [inline lib/utilities/extensions.dart:5]: some strings are nullable when you call `captilize` also add null checks here

**Review by dbenedictos — CHANGES_REQUESTED** [review]

File: lib/utilities/extensions.dart (hunk 15 above)
Comment (dbenedictos) [reply lib/utilities/extensions.dart:5]: ahh no, never mind

**Review by dbenedictos — CHANGES_REQUESTED** [review]

File: lib/feature/widgets/type_container_widget.dart [hunk 16]
```diff
@@ -0,0 +1,32 @@
+import 'package:flutter/material.dart';
+import 'package:pokedex_async_redux/utilities/constants.dart';
+import 'package:pokedex_async_redux/utilities/extensions.dart';
+
+class TypeWidget extends StatelessWidget {
+  const TypeWidget({
+    required this.type,
```
Comment (dbenedictos) [inline lib/feature/widgets/type_container_widget.dart]: trailing comma

File: lib/feature/pokemon_details_page.dart/widgets/pokemon_moves_tab.dart [hunk 17]
```diff
@@ -0,0 +1,38 @@
...
+  final List<Moves>? moves;
+  final Color color;
+  @override
+  Widget build(BuildContext context) {
+    return SingleChildScrollView(
+      child: Wrap(
+        children: moves
+                ?.map((e) => Padding(
+                      padding: EdgeInsets.all(2),
+                      child: Container(
```
Comment (dbenedictos) [inline lib/feature/pokemon_details_page.dart/widgets/pokemon_moves_tab.dart]: use your TypeContainerWidget here

**Review by dbenedictos — CHANGES_REQUESTED** [review]

File: lib/feature/widgets/type_container_widget.dart [hunk 18]
```diff
@@ -0,0 +1,32 @@
+import 'package:flutter/material.dart';
+import 'package:pokedex_async_redux/utilities/constants.dart';
+import 'package:pokedex_async_redux/utilities/extensions.dart';
+
+class TypeWidget extends StatelessWidget {
```
Comment (dbenedictos) [inline lib/feature/widgets/type_container_widget.dart]: match your widget name to your file name. or vice-versa

**Review by dbenedictos — CHANGES_REQUESTED** [review]

File: lib/feature/home_page/widgets/pokemon_tile.dart (hunk 11 above)
Comment (dbenedictos) [inline lib/feature/home_page/widgets/pokemon_tile.dart]: async

File: lib/feature/home_page/widgets/pokemon_tile.dart [hunk 19]
```diff
@@ -7,26 +7,25 @@ import 'package:pokedex_async_redux/feature/pokemon_details_page.dart/pokemon_de
...
+  PokemonType? thisTileTypes = PokemonType();
   @override
   void initState() {
-    http.get(Uri.tryParse('${widget.thisPokemon.url}') ?? Uri()).then((value) {
-      if (value.statusCode == 200) {
-        var result = jsonDecode(value.body);
-        List SubTypeMap = result['types'];
-        setState(() {
+    WidgetsBinding.instance.addPostFrameCallback((_) {
+      http.get(Uri.tryParse('${widget.thisPokemon.url}') ?? Uri()).then((value) {
```
Comment (dbenedictos) [inline lib/feature/home_page/widgets/pokemon_tile.dart]: await

File: lib/feature/home_page/widgets/pokemon_tile.dart [hunk 20]
```diff
@@ -36,14 +35,22 @@ class _PokemonTileState extends State<PokemonTile> {
                   url: e['type']['url'],
                 ))).toList(),
           );
-        });
-      } else {
-        print('error');
-      }
+          if (mounted) {
```
Comment (dbenedictos) [inline lib/feature/home_page/widgets/pokemon_tile.dart]: put this if statement after the outside the `then` method

File: lib/feature/home_page/widgets/pokemon_tile.dart [hunk 21]
```diff
@@ -36,14 +35,22 @@ class _PokemonTileState extends State<PokemonTile> {
...
+        } else {
+          print('error');
+        }
+      });
     });
     super.initState();
   }
 
+  @override
+  void dispose() {
```
Comment (dbenedictos) [inline lib/feature/home_page/widgets/pokemon_tile.dart]: no need for the dispose override. because you're not doing anything inside of it

**Review by dbenedictos — CHANGES_REQUESTED** [review]
Last set of comments

**Pull request merged**